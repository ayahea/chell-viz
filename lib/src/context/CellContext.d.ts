/// <reference types="react" />
import { CELL_TYPE } from 'chell';
import * as React from 'react';
export interface ICellContext {
  addCells: (cells: CELL_TYPE[]) => void;
  currentCells: CELL_TYPE[];
  removeAllCells: (cells: CELL_TYPE[]) => void;
  removeCells: (cells: CELL_TYPE[]) => void;
}
export declare const initialCellContext: ICellContext;
export declare const CellContext: React.Context<{
  addCells: (cells: number[]) => void;
  currentCells: number[];
  removeAllCells: (cells: number[]) => void;
  removeCells: (cells: number[]) => void;
}>;
