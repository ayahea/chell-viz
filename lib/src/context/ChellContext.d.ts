/// <reference types="react" />
import * as React from 'react';
import * as ResContext from '../context/ResidueContext';
export declare const initialState: {
  cellContext: {
    addCells: (cells: number[]) => void;
    currentCells: number[];
    removeAllCells: (cells: number[]) => void;
    removeCells: (cells: number[]) => void;
  };
  residueContext: {
    addCandidateResidues: (residues: number[]) => void;
    addHoveredResidues: (residues: number[]) => void;
    addLockedResiduePair: (residues: number[]) => void;
    candidateResidues: number[];
    hoveredResidues: number[];
    lockedResiduePairs: ResContext.IResidueSelection;
    removeAllLockedResiduePairs: () => void;
    removeCandidateResidues: () => void;
    removeHoveredResidues: () => void;
    removeLockedResiduePair: (residues: number[]) => void;
  };
};
export declare type ChellContextState = Readonly<typeof initialState>;
export declare class ChellContext extends React.Component<any, ChellContextState> {
  public readonly state: ChellContextState;
  constructor(props: any);
  public render(): JSX.Element;
  public onAddCells: (cells: number[]) => void;
  public onRemoveAllCells: () => void;
  public onRemoveCells: (cellsToRemove: number[]) => void;
  public onAddCandidateResidues: (candidateResidues: number[]) => void;
  public onAddHoveredResidues: (hoveredResidues: number[]) => void;
  public onRemoveAllResidues: () => void;
  public onRemoveCandidateResidue: () => void;
  public onRemoveHoveredResidue: () => void;
  public onRemoveResidues: (residues: number[]) => void;
  public onResidueSelect: (residues: number[]) => void;
}
