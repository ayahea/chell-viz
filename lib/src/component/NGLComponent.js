"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var NGL = require("ngl");
var React = require("react");
var semantic_ui_react_1 = require("semantic-ui-react");
var ResidueContext_1 = require("../context/ResidueContext");
var ReactHelper_1 = require("../helper/ReactHelper");
exports.SUPPORTED_REPS = [
    'axes',
    'backbone',
    'ball+stick',
    'distance',
    'label',
    'line',
    'hyperball',
    'spacefill',
];
var defaultProps = __assign({ data: undefined, height: 400 }, ResidueContext_1.initialResidueContext, { padding: 0, width: 400 });
var initialState = {
    max_x: 0,
    min_x: 1000,
    nodeSize: 4,
    probabilityFilter: 0.99,
    residueOffset: 0,
    stage: undefined,
    structureComponent: undefined,
};
exports.NGLComponentWithDefaultProps = ReactHelper_1.withDefaultProps(defaultProps, /** @class */ (function (_super) {
    __extends(NGLComponentClass, _super);
    function NGLComponentClass(props) {
        var _this = _super.call(this, props) || this;
        _this.state = initialState;
        _this.canvas = null;
        _this.residueSelectionRepresentations = {};
        _this.onClick = function (pickingProxy) {
            var _a = _this.props, addCandidateResidues = _a.addCandidateResidues, addLockedResiduePair = _a.addLockedResiduePair, candidateResidues = _a.candidateResidues, removeCandidateResidues = _a.removeCandidateResidues, removeHoveredResidues = _a.removeHoveredResidues;
            var structureComponent = _this.state.structureComponent;
            if (pickingProxy && (pickingProxy.atom || pickingProxy.bond)) {
                var atom = pickingProxy.atom || pickingProxy.closestBondAtom;
                var resno = atom.resno + _this.state.residueOffset;
                if (candidateResidues.length >= 1) {
                    addLockedResiduePair(candidateResidues.concat([resno]));
                    removeCandidateResidues();
                }
                else {
                    addCandidateResidues([resno]);
                }
            }
            else if (structureComponent) {
                // User clicked off-structure, so clear non-locked residue state.
                _this.removeNonLockedRepresentations(structureComponent);
                removeCandidateResidues();
                removeHoveredResidues();
            }
        };
        return _this;
    }
    NGLComponentClass.prototype.componentDidMount = function () {
        if (this.canvas) {
            var stage = new NGL.Stage(this.canvas);
            this.setState({
                stage: stage,
            });
            var data = this.props.data;
            if (data) {
                this.addStructureToStage(data, stage);
            }
        }
    };
    NGLComponentClass.prototype.componentWillUnmount = function () {
        var stage = this.state.stage;
        if (stage) {
            stage.dispose();
            this.setState({
                stage: undefined,
            });
        }
    };
    NGLComponentClass.prototype.componentDidUpdate = function (prevProps, prevState) {
        var _this = this;
        var _a = this.props, data = _a.data, lockedResiduePairs = _a.lockedResiduePairs;
        var _b = this.state, stage = _b.stage, structureComponent = _b.structureComponent;
        var isNewData = data && data !== prevProps.data;
        if (data && stage && isNewData) {
            stage.removeAllComponents();
            this.addStructureToStage(data, stage);
        }
        if (structureComponent && prevProps.lockedResiduePairs !== lockedResiduePairs) {
            this.removeHighlights(structureComponent, prevProps.lockedResiduePairs);
            Object.keys(lockedResiduePairs).forEach(function (key) {
                _this.highlightElement(structureComponent, lockedResiduePairs[key]);
            });
        }
        if (structureComponent && this.props.hoveredResidues !== prevProps.hoveredResidues) {
            this.removeNonLockedRepresentations(structureComponent);
            this.highlightElement(structureComponent, this.props.hoveredResidues);
        }
    };
    /**
     * Renders the NGL canvas.
     *
     * Because we are working with WebGL via the canvas, updating this visualization happens through the canvas reference.
     *
     * @returns The NGL Component
     */
    NGLComponentClass.prototype.render = function () {
        var _this = this;
        var _a = this.props, height = _a.height, padding = _a.padding, width = _a.width;
        return (React.createElement("div", { id: "NGLComponent", style: { padding: padding } },
            React.createElement("div", { ref: function (el) { return (_this.canvas = el); }, style: { height: height, width: width } }),
            React.createElement(semantic_ui_react_1.GridRow, null,
                React.createElement(semantic_ui_react_1.Button, { onClick: this.props.removeAllLockedResiduePairs }, "Remove Locked Residues"))));
    };
    /**
     * Adds a NGL structure to the stage.
     *
     * @param data A NGL Structure.
     * @param stage A NGL Stage.
     */
    NGLComponentClass.prototype.addStructureToStage = function (data, stage) {
        var _this = this;
        var structureComponent = stage.addComponentFromObject(data);
        this.setState({
            residueOffset: data.residueStore.resno[0],
            structureComponent: structureComponent,
        });
        stage.defaultFileRepresentation(structureComponent);
        structureComponent.reprList.forEach(function (rep) {
            rep.setParameters({ opacity: 1.0 });
        });
        structureComponent.stage.mouseControls.add("hoverPick" /* HOVER_PICK */, function (aStage, pickingProxy) { return _this.onHover(aStage, pickingProxy, data, structureComponent); });
        stage.signals.clicked.add(this.onClick);
    };
    NGLComponentClass.prototype.onHover = function (aStage, pickingProxy, data, structureComponent) {
        if (pickingProxy && (pickingProxy.atom || pickingProxy.bond)) {
            var atom = pickingProxy.atom || pickingProxy.closestBondAtom;
            var resno = atom.resno + this.state.residueOffset;
            this.removeNonLockedRepresentations(structureComponent);
            this.highlightElement(structureComponent, [resno]);
            this.props.addHoveredResidues([resno]);
            var candidateResidues = this.props.candidateResidues;
            if (candidateResidues.length >= 1) {
                this.highlightElement(structureComponent, candidateResidues.concat([resno]));
            }
        }
    };
    NGLComponentClass.prototype.removeHighlights = function (structureComponent, residues) {
        var repDict = this.residueSelectionRepresentations;
        Object.keys(residues).forEach(function (prevKey) {
            repDict[prevKey].map(function (rep) { return structureComponent.removeRepresentation(rep); });
        });
    };
    /**
     * Highlight a specific residue on a 3D structure.
     *
     * @param structureComponent The structure for which the residue to highlight belongs.
     * @param selection [NGL Selection](https://github.com/arose/ngl/blob/master/doc/usage/selection-language.md) for what to highlight.
     */
    NGLComponentClass.prototype.highlightElement = function (structureComponent, residues) {
        var residueOffset = this.state.residueOffset;
        var residueKey = residues.toString();
        var residueWithOffset = residues.map(function (res) { return res - residueOffset; });
        var repDict = this.residueSelectionRepresentations;
        if (repDict[residueKey]) {
            repDict[residueKey].map(function (rep) { return structureComponent.removeRepresentation(rep); });
        }
        else {
            repDict[residueKey] = [];
        }
        var selection = residueWithOffset.join('.CA, ') + '.CA';
        if (residueWithOffset.length >= 2) {
            repDict[residueKey].push(structureComponent.addRepresentation('distance', {
                atomPair: [selection.split(',')],
                color: 'skyblue',
                labelUnit: 'nm',
            }));
        }
        if (residueWithOffset.length !== 0) {
            repDict[residueKey].push(structureComponent.addRepresentation('ball+stick', {
                sele: residueWithOffset.join(', '),
            }));
        }
    };
    NGLComponentClass.prototype.removeNonLockedRepresentations = function (structureComponent) {
        var repDict = this.residueSelectionRepresentations;
        for (var _i = 0, _a = Object.keys(repDict); _i < _a.length; _i++) {
            var key = _a[_i];
            if (!this.props.lockedResiduePairs[key]) {
                repDict[key].forEach(function (rep) { return structureComponent.removeRepresentation(rep); });
                delete this.residueSelectionRepresentations[key];
            }
        }
    };
    return NGLComponentClass;
}(React.Component)));
exports.NGLComponent = function (props) { return (React.createElement(ResidueContext_1.ResidueContext.Consumer, null, function (_a) {
    var addLockedResiduePair = _a.addLockedResiduePair, addHoveredResidues = _a.addHoveredResidues, addCandidateResidues = _a.addCandidateResidues, candidateResidues = _a.candidateResidues, hoveredResidues = _a.hoveredResidues, lockedResiduePairs = _a.lockedResiduePairs, removeAllLockedResiduePairs = _a.removeAllLockedResiduePairs, removeCandidateResidues = _a.removeCandidateResidues, removeHoveredResidues = _a.removeHoveredResidues;
    return (React.createElement(exports.NGLComponentWithDefaultProps, __assign({}, props, { addCandidateResidues: addCandidateResidues, addHoveredResidues: addHoveredResidues, addLockedResiduePair: addLockedResiduePair, candidateResidues: candidateResidues, hoveredResidues: hoveredResidues, lockedResiduePairs: lockedResiduePairs, removeAllLockedResiduePairs: removeAllLockedResiduePairs, removeCandidateResidues: removeCandidateResidues, removeHoveredResidues: removeHoveredResidues })));
})); };
