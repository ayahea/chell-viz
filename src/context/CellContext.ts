import { CELL_TYPE } from 'chell';
import * as React from 'react';

export const initialCellContext = {
  addCells: (cells: CELL_TYPE[]) => {
    return;
  },
  currentCells: [] as CELL_TYPE[],
  removeAllCells: (cells: CELL_TYPE[]) => {
    return;
  },
  removeCells: (cells: CELL_TYPE[]) => {
    return;
  },
};

export const CellContext = React.createContext({
  ...initialCellContext,
});
