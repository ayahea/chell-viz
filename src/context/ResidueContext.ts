import { RESIDUE_TYPE } from 'chell';
import * as React from 'react';

export interface IResidueSelection {
  [key: string]: RESIDUE_TYPE[];
}

export interface IResidueContext {
  addCandidateResidue: (residue: RESIDUE_TYPE) => void;
  addLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
  candidateResidue: RESIDUE_TYPE | 'none';
  lockedResiduePairs: IResidueSelection;
  removeAllLockedResiduePairs: () => void;
  removeCandidateResidue: () => void;
  removeLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
}

export const initialResidueContext = {
  candidateResidue: 'none' as RESIDUE_TYPE | 'none',
  lockedResiduePairs: {} as IResidueSelection,
} as IResidueContext;

export const ResidueContext = React.createContext({
  ...initialResidueContext,
});
