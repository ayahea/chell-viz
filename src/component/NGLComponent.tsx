import * as NGL from 'ngl';
import { PickingProxy, Stage, StructureComponent, StructureRepresentationType } from 'ngl';
import * as React from 'react';
import { Button, GridRow } from 'semantic-ui-react';

import { initialResidueContext, IResidueSelection, ResidueContext } from '../context/ResidueContext';
import { withDefaultProps } from '../helper/ReactHelper';

export type NGL_HOVER_CB_RESULT_TYPE = number;

export interface IRepresentationDict {
  [key: string]: NGL.RepresentationElement[];
}

export const SUPPORTED_REPS: StructureRepresentationType[] = [
  'axes',
  'backbone',
  'ball+stick',
  'distance',
  'label',
  'line',
  'hyperball',
  'spacefill',
];

const defaultProps = {
  data: undefined as NGL.Structure | undefined,
  height: 400,
  ...initialResidueContext,
  padding: 0,
  width: 400,
};

const initialState = {
  max_x: 0,
  min_x: 1000,
  nodeSize: 4,
  probabilityFilter: 0.99,
  residueOffset: 0,
  stage: undefined as NGL.Stage | undefined,
  structureComponent: undefined as NGL.StructureComponent | undefined,
};

type Props = {} & typeof defaultProps;
type State = Readonly<typeof initialState>;

export const NGLComponentWithDefaultProps = withDefaultProps(
  defaultProps,
  class NGLComponentClass extends React.Component<Props, State> {
    public readonly state: State = initialState;

    protected canvas: HTMLElement | null = null;
    protected residueSelectionRepresentations: IRepresentationDict = {};

    constructor(props: any) {
      super(props);
    }

    public componentDidMount() {
      if (this.canvas) {
        const stage = new NGL.Stage(this.canvas);

        this.setState({
          stage,
        });

        const { data } = this.props;
        if (data) {
          this.addStructureToStage(data, stage);
        }
      }
    }

    public componentWillUnmount() {
      const { stage } = this.state;
      if (stage) {
        stage.dispose();
        this.setState({
          stage: undefined,
        });
      }
    }

    public componentDidUpdate(prevProps: Props, prevState: State) {
      const { data, lockedResiduePairs } = this.props;
      const { stage, structureComponent } = this.state;

      const isNewData = data && data !== prevProps.data;
      if (data && stage && isNewData) {
        stage.removeAllComponents();
        this.addStructureToStage(data, stage);
      }

      if (structureComponent && prevProps.lockedResiduePairs !== lockedResiduePairs) {
        this.removeHighlights(structureComponent, prevProps.lockedResiduePairs);
        Object.keys(lockedResiduePairs).forEach(key => {
          this.highlightElement(structureComponent, lockedResiduePairs[key]);
        });
      }

      if (
        structureComponent &&
        this.props.hoveredResidue !== 'none' &&
        this.props.hoveredResidue !== prevProps.hoveredResidue
      ) {
        this.removeNonLockedRepresentations(structureComponent);
        this.highlightElement(structureComponent, [this.props.hoveredResidue]);
      }
    }

    /**
     * Renders the NGL canvas.
     *
     * Because we are working with WebGL via the canvas, updating this visualization happens through the canvas reference.
     *
     * @returns The NGL Component
     */
    public render() {
      const { height, padding, width } = this.props;
      return (
        <div id="NGLComponent" style={{ padding }}>
          <div ref={el => (this.canvas = el)} style={{ height, width }} />
          <GridRow>
            <Button onClick={this.props.removeAllLockedResiduePairs}>Remove Locked Residues</Button>
          </GridRow>
        </div>
      );
    }

    /**
     * Adds a NGL structure to the stage.
     *
     * @param data A NGL Structure.
     * @param stage A NGL Stage.
     */
    protected addStructureToStage(data: NGL.Structure, stage: NGL.Stage) {
      const structureComponent = stage.addComponentFromObject(data) as NGL.StructureComponent;

      this.setState({
        residueOffset: data.residueStore.resno[0],
        structureComponent,
      });

      stage.defaultFileRepresentation(structureComponent);
      structureComponent.reprList.forEach(rep => {
        rep.setParameters({ opacity: 1.0 });
      });

      structureComponent.stage.mouseControls.add(
        NGL.MouseActions.HOVER_PICK,
        (aStage: Stage, pickingProxy: PickingProxy) => this.onHover(aStage, pickingProxy, data, structureComponent),
      );

      stage.signals.clicked.add(this.onClick);
    }

    protected onHover(
      aStage: Stage,
      pickingProxy: PickingProxy,
      data: NGL.Structure,
      structureComponent: StructureComponent,
    ) {
      if (pickingProxy && (pickingProxy.atom || pickingProxy.bond)) {
        const atom = pickingProxy.atom || pickingProxy.closestBondAtom;
        const resno = atom.resno + this.state.residueOffset;

        this.removeNonLockedRepresentations(structureComponent);

        this.highlightElement(structureComponent, [resno]);
        this.props.addHoveredResidue(resno);

        const { candidateResidue } = this.props;
        if (candidateResidue !== 'none') {
          this.highlightElement(structureComponent, [candidateResidue, resno]);
        }
      }
    }

    protected onClick = (pickingProxy: PickingProxy) => {
      const {
        addCandidateResidue,
        addLockedResiduePair,
        candidateResidue,
        removeCandidateResidue,
        removeHoveredResidue,
      } = this.props;
      const { structureComponent } = this.state;

      if (pickingProxy && (pickingProxy.atom || pickingProxy.bond)) {
        const atom = pickingProxy.atom || pickingProxy.closestBondAtom;
        const resno = atom.resno + this.state.residueOffset;

        if (candidateResidue !== 'none') {
          addLockedResiduePair([candidateResidue, resno]);
          removeCandidateResidue();
        } else {
          addCandidateResidue(resno);
        }
      } else if (structureComponent) {
        // User clicked off-structure, so clear non-locked residue state.
        this.removeNonLockedRepresentations(structureComponent);
        removeCandidateResidue();
        removeHoveredResidue();
      }
    };

    protected removeHighlights(structureComponent: StructureComponent, residues: IResidueSelection) {
      const repDict = this.residueSelectionRepresentations;
      Object.keys(residues).forEach(prevKey => {
        repDict[prevKey].map(rep => structureComponent.removeRepresentation(rep));
      });
    }

    /**
     * Highlight a specific residue on a 3D structure.
     *
     * @param structureComponent The structure for which the residue to highlight belongs.
     * @param selection [NGL Selection](https://github.com/arose/ngl/blob/master/doc/usage/selection-language.md) for what to highlight.
     */
    protected highlightElement(structureComponent: StructureComponent, residues: number[]) {
      const { residueOffset } = this.state;
      const residueKey = residues.toString();
      const residueWithOffset = residues.map(res => res - residueOffset);
      const repDict = this.residueSelectionRepresentations;

      if (repDict[residueKey]) {
        repDict[residueKey].map(rep => structureComponent.removeRepresentation(rep));
      } else {
        repDict[residueKey] = [];
      }

      const selection = residueWithOffset.join('.CA, ') + '.CA';
      if (residueWithOffset.length >= 2) {
        repDict[residueKey].push(
          structureComponent.addRepresentation('distance', {
            atomPair: [selection.split(',')],
            color: 'skyblue',
            labelUnit: 'nm',
          }),
        );
      }

      repDict[residueKey].push(
        structureComponent.addRepresentation('ball+stick', {
          sele: residueWithOffset.join(', '),
        }),
      );
    }

    protected removeNonLockedRepresentations(structureComponent: NGL.StructureComponent) {
      const repDict = this.residueSelectionRepresentations;
      for (const key of Object.keys(repDict)) {
        if (!this.props.lockedResiduePairs[key]) {
          repDict[key].forEach(rep => structureComponent.removeRepresentation(rep));
          delete this.residueSelectionRepresentations[key];
        }
      }
    }
  },
);

// TODO The required props should be discernable from `withDefaultProps` without needing to duplicate.
// However the Context consumer syntax is still new to me and I can't find the right combination :(
type requiredProps = Partial<typeof defaultProps> & Required<Omit<Props, keyof typeof defaultProps>>;

export const NGLComponent = (props: requiredProps) => (
  <ResidueContext.Consumer>
    {({
      addLockedResiduePair,
      addHoveredResidue,
      addCandidateResidue,
      candidateResidue,
      hoveredResidue,
      lockedResiduePairs,
      removeAllLockedResiduePairs,
      removeCandidateResidue,
      removeHoveredResidue,
    }) => (
      <NGLComponentWithDefaultProps
        {...props}
        addCandidateResidue={addCandidateResidue}
        addHoveredResidue={addHoveredResidue}
        addLockedResiduePair={addLockedResiduePair}
        candidateResidue={candidateResidue}
        hoveredResidue={hoveredResidue}
        lockedResiduePairs={lockedResiduePairs}
        removeAllLockedResiduePairs={removeAllLockedResiduePairs}
        removeCandidateResidue={removeCandidateResidue}
        removeHoveredResidue={removeHoveredResidue}
      />
    )}
  </ResidueContext.Consumer>
);
