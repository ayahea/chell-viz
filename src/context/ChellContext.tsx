import * as React from 'react';

import { CELL_TYPE, RESIDUE_TYPE } from '../../types/chell';
import CellContext, { initialCellContext } from '../context/CellContext';
import ResidueContext, { initialResidueContext } from '../context/ResidueContext';

export const initialState = {
  cellContext: {
    ...initialCellContext,
  },
  residueContext: {
    ...initialResidueContext,
  },
};

export type ChellContextState = Readonly<typeof initialState>;

export default class ChellContext extends React.Component<any, ChellContextState> {
  public readonly state: ChellContextState = initialState;

  constructor(props: any) {
    super(props);
    this.state = {
      ...this.state,
      cellContext: {
        ...this.state.cellContext,
        addCells: this.onAddCells,
        removeAllCells: this.onRemoveAllCells,
        removeCells: this.onRemoveCells,
      },
      residueContext: {
        ...this.state.residueContext,
        addCandidateResidues: this.onAddCandidateResidues,
        addHoveredResidues: this.onAddHoveredResidues,
        addLockedResiduePair: this.onResidueSelect,
        removeAllLockedResiduePairs: this.onRemoveAllResidues,
        removeCandidateResidues: this.onRemoveCandidateResidue,
        removeHoveredResidues: this.onRemoveHoveredResidue,
        removeLockedResiduePair: this.onRemoveResidues,
      },
    };
  }

  public render() {
    return (
      <CellContext.Provider value={this.state.cellContext}>
        <ResidueContext.Provider value={this.state.residueContext}>{this.props.children}</ResidueContext.Provider>
      </CellContext.Provider>
    );
  }

  public onAddCells = (cells: CELL_TYPE[]) => {
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: cells,
      },
    });
  };

  public onRemoveAllCells = () => {
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: [],
      },
    });
  };

  public onRemoveCells = (cellsToRemove: CELL_TYPE[]) => {
    const { currentCells } = this.state.cellContext;
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: currentCells.filter(cell => cellsToRemove.indexOf(cell) === -1),
      },
    });
  };

  public onAddCandidateResidues = (candidateResidues: RESIDUE_TYPE[]) => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: candidateResidues.sort(),
      },
    });
  };

  public onAddHoveredResidues = (hoveredResidues: RESIDUE_TYPE[]) => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        hoveredResidues: hoveredResidues.sort(),
      },
    });
  };

  public onRemoveAllResidues = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        lockedResiduePairs: {},
      },
    });
  };

  public onRemoveCandidateResidue = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: [],
      },
    });
  };

  public onRemoveHoveredResidue = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        hoveredResidues: [],
      },
    });
  };

  public onRemoveResidues = (residues: RESIDUE_TYPE[]) => {
    const residueKey = residues.join(',');
    const { lockedResiduePairs } = this.state.residueContext;
    if (lockedResiduePairs[residueKey]) {
      delete lockedResiduePairs[residueKey];
    }
  };

  public onResidueSelect = (residues: RESIDUE_TYPE[]) => {
    const { lockedResiduePairs } = this.state.residueContext;
    const sortedResidues = residues.sort();
    const residuePairKey = sortedResidues.toString();
    if (!lockedResiduePairs[residuePairKey]) {
      this.setState({
        residueContext: {
          ...this.state.residueContext,
          lockedResiduePairs: {
            ...lockedResiduePairs,
            [residuePairKey]: sortedResidues,
          },
        },
      });
    }
  };
}
