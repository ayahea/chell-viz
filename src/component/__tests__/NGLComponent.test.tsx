import { mount } from 'enzyme';
import * as React from 'react';
import * as Renderer from 'react-test-renderer';

import toJson from 'enzyme-to-json';
import { initialResidueContext, IResidueContext } from '../../context/ResidueContext';
import NGLComponent from '../NGLComponent';

jest.mock('ngl');

beforeEach(() => {
  jest.resetModules();
});

// https://medium.com/@ryandrewjohnson/unit-testing-components-using-reacts-new-context-api-4a5219f4b3fe
// Provides a dummy context for unit testing purposes.
const getComponentWithContext = (context: IResidueContext = { ...initialResidueContext }) => {
  jest.doMock('../../context/ResidueContext', () => {
    return {
      ResidueContext: {
        Consumer: (props: any) => props.children(context),
      },
    };
  });

  return require('../NGLComponent');
};

describe('NGLComponent', () => {
  const sampleData = {
    residueStore: {
      resno: [0, 1, 2, 3, 4],
    },
  };

  test("Should match existing snapshot when canvas isn't available.", () => {
    expect(Renderer.create(<NGLComponent />).toJSON()).toMatchSnapshot();
  });

  test('Should match existing snapshot when given sample data', () => {
    const Component = getComponentWithContext();
    const wrapper = mount(<Component.NGLComponentClass data={sampleData} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should handle prop updates.', () => {
    const Component = getComponentWithContext();
    const wrapper = mount(<Component.NGLComponentClass />);
    const initialProps = wrapper.props();

    wrapper.setProps({
      hoveredResidues: [0],
    });

    expect(wrapper.props()).not.toEqual(initialProps);
  });

  test('Should handle data updates.', () => {
    const Component = getComponentWithContext();
    const wrapper = mount(<Component.NGLComponentClass />);

    wrapper.setProps({
      data: sampleData,
    });
  });

  test('Should unmount correctly.', () => {
    const Component = getComponentWithContext();
    const wrapper = mount(<Component.NGLComponentClass />);
    expect(wrapper.get(0)).not.toBeNull();
    wrapper.unmount();
    expect(wrapper.get(0)).toBeNull();
  });
});
