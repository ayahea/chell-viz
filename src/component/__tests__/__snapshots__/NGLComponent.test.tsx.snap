// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NGLComponent Should match existing snapshot when canvas isn't available. 1`] = `
<div
  id="NGLComponent"
  style={
    Object {
      "padding": 0,
    }
  }
>
  <div
    style={
      Object {
        "height": 400,
        "width": 400,
      }
    }
  />
  <div
    className="row"
  >
    <button
      className="ui button"
      disabled={undefined}
      onClick={[Function]}
      role="button"
      tabIndex={undefined}
    >
      Remove Locked Residues
    </button>
  </div>
</div>
`;

exports[`NGLComponent Should match existing snapshot when given sample data 1`] = `
<NGLComponentClass
  data={
    Object {
      "residueStore": Object {
        "resno": Array [
          0,
          1,
          2,
          3,
          4,
        ],
      },
    }
  }
  height={400}
  padding={0}
  width={400}
>
  <div
    id="NGLComponent"
    style={
      Object {
        "padding": 0,
      }
    }
  >
    <div
      style={
        Object {
          "height": 400,
          "width": 400,
        }
      }
    />
    <GridRow>
      <div
        className="row"
      >
        <Button
          as="button"
          role="button"
        >
          <button
            className="ui button"
            onClick={[Function]}
            role="button"
          >
            Remove Locked Residues
          </button>
        </Button>
      </div>
    </GridRow>
  </div>
</NGLComponentClass>
`;
